lib/resty/ldap/filter.lua

```diff
@@ -111,19 +115,32 @@ local filter = P{
     ATTRIBUTE_VALUE = rawValue / function(value)
         return { attribute_value = value }
     end,
-    ATTRIBUTE_VALUE_SUBSTRING =
-        ((V'ATTRIBUTE_VALUE' * V'WILDCARD') +
-        (V'WILDCARD' * V'ATTRIBUTE_VALUE' * V'WILDCARD') +
-        (V'WILDCARD' * V'ATTRIBUTE_VALUE')) / function(...)
-            local s = {}
-            for i = 1, select('#', ...) do
-                local v = select(i, ...)
-                table_insert(s, type(v) == "table" and v.attribute_value or v)
+    ATTRIBUTE_VALUE_SUBSTRING = (
+        maybe('ATTRIBUTE_VALUE') *
+        V'WILDCARD' *
+        list(V'ATTRIBUTE_VALUE' * V'WILDCARD', 0) *
+        maybe('ATTRIBUTE_VALUE')
+    ) / function(...)
+        local s = {}
+        for i = 1, select('#', ...) do
+            local v = select(i, ...)
+            if v.attribute_value then
+                table_insert(s, v.attribute_value)
+            elseif #v > 0 and v[1] then -- check if the elements are arrays
+                -- When there is the following format xxx*a*b*c*xxx,
+                -- the parser will output a*b*c* as a nested two-dimensional array,
+                -- which we need to flatten
+                for _, any_item in ipairs(v) do
```

Comment:
Do we need to check if any_item.attribute_value is not nil?

Reply:
Hi, @spacewander.

This table is generated directly by the parser, which cannot generate a table that does not contain `attribute_value`.

In principle, each element contained in P{} is a matching rule, which may contain some other sub-level matching rules, and the parent level contains all the possible output values of those sub-levels, so the values that will not be output from the end of this "syntax tree" will never be captured by the parent node.

According to the matching rules you can find that `ATTRIBUTE_VALUE_SUBSTRING` consists of two sub-rules, `ATTRIBUTE_VALUE` and `WILDCARD`, which both determine the output `attribute_value` field, so `ATTRIBUTE_VALUE_SUBSTRING`, as the parent, will not catch tables containing nil values.

Reply:
When an exception occurs, the most likely scenario is that the compiler outputs a syntax error hint, rather than a Lua error during parsing.